napari/_qt/qt_dims_slider.py

```diff
@@ -0,0 +1,471 @@
+    fps_changed = Signal(float)
+    mode_changed = Signal(int)
+    range_changed = Signal(tuple)
+    play_started = Signal()
+    play_stopped = Signal()
+
+    def __init__(self, parent: QWidget, axis: int):
+        super().__init__(parent=parent)
+        self.axis = axis
+        self.qt_dims = parent
+        self.dims = parent.dims
+        self.label = None
+        self.slider = None
+        self.play_button = None
+
+        self._fps = 10
+        self._minframe = None
+        self._maxframe = None
+        self._loop_mode = 1
+
+        layout = QHBoxLayout()
+        self._create_axis_label_widget()
+        self._create_range_slider_widget()
+        self._create_play_button_widget()
+        layout.addWidget(self.label)
+        layout.addWidget(self.play_button)
+        layout.addWidget(self.slider, stretch=1)
+        layout.setContentsMargins(0, 0, 0, 0)
+        layout.setSpacing(3)
+        self.setLayout(layout)
+        self.dims.events.axis_labels.connect(self._pull_label)
+
+    def _create_play_button_widget(self):
+        """Creates the actual play button, which has the modal popup."""
+        self.play_button = QtPlayButton(self.qt_dims, self.axis)
+        self.play_button.mode_combo.currentIndexChanged.connect(
+            lambda x: self.__class__.loop_mode.fset(self, x)
+        )
+
+        def fps_listener(*args):
+            fps = self.play_button.fpsspin.value()
+            fps *= -1 if self.play_button.reverse_check.isChecked() else 1
+            self.__class__.fps.fset(self, fps)
+
+        self.play_button.fpsspin.editingFinished.connect(fps_listener)
+        self.play_button.reverse_check.stateChanged.connect(fps_listener)
+        self.play_stopped.connect(self.play_button._handle_stop)
+        self.play_started.connect(self.play_button._handle_start)
+
+    def _pull_label(self, event):
+        """updates the label LineEdit from the dims model"""
+        if event.axis == self.axis:
+            label = self.dims.axis_labels[self.axis]
+            self.label.setText(label)
+            self.label_changed.emit(self.axis, label)
+
+    def _update_label(self):
+        with self.dims.events.axis_labels.blocker():
+            self.dims.set_axis_label(self.axis, self.label.text())
+        self.label.clearFocus()
+        self.qt_dims.setFocus()
+        self.label_changed.emit(self.axis, self.label.text())
+
+    def _create_axis_label_widget(self):
+        """Create the axis label widget which accompanies its slider."""
+        label = QLineEdit(self)
+        label.setObjectName('axis_label')  # needed for _update_label
+        label.setText(self.dims.axis_labels[self.axis])
+        label.home(False)
+        label.setToolTip('Edit to change axis label')
+        label.setAcceptDrops(False)
+        label.setEnabled(True)
+        label.setAlignment(Qt.AlignRight)
+        label.setContentsMargins(0, 0, 2, 0)
+        label.editingFinished.connect(self._update_label)
```

Comment:
Curious your thoughts on splitting out the `editingFinished` callbacks which we just want triggered once at the end and also a `textEdited` callback that does the resizing so the resizing will happen as you type. This could be nice as right now if i start typing something long it actually disappears part way through my typing and is only revealed at the end

Reply:
could do.  I avoided it at first to prevent a ton of redraws getting sent, but it's not like they'll happen unless someone is actively editing. so might work nicely.  Did you try it?  I'll try and change if it feels better.

Reply:
Just tried it - it feels great. I'd just swap `textEdited` for `editingFinished` here and then move
```python
        self.label.clearFocus()
        self.qt_dims.setFocus()
```
to its own function and connect that on `editingFinished` or just drop them entirely. I like them though as they get rid of the flashing cursor when you are done.

Reply:
yeah, I like it!